scene/gui/popup_button.cpp

```diff
@@ -0,0 +1,154 @@
+/*************************************************************************/
+/*  popup_button.cpp                                                     */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "popup_button.h"
+
+#include "core/os/keyboard.h"
+#include "scene/main/window.h"
+
+void PopupButton::set_popup(Popup *p_popup_node) {
+	ERR_FAIL_COND(!p_popup_node);
+	ERR_FAIL_COND_MSG(p_popup_node->get_parent() != this, "Connected popup must be a child of the button.");
+
+	if (popup) {
+		popup->disconnect("about_to_popup", callable_mp(this, &PopupButton::_popup_visibility_changed));
+		popup->disconnect("popup_hide", callable_mp(this, &PopupButton::_popup_visibility_changed));
+	}
```

Comment:
I think replacing the popup is a bit awkward. You have to free it yourself, but you can't do it before setting the new popup. Maybe this should also automatically free the old popup?